include/nbl/builtin/hlsl/random/pcg.hlsl

```diff
@@ -0,0 +1,62 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#ifndef _NBL_BUILTIN_GLSL_RANDOM_PCG_HLSL_INCLUDED_
+#define _NBL_BUILTIN_GLSL_RANDOM_PCG_HLSL_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+
+struct PCGStateHolder
+{
+    void pcg32_state_advance()
+    {
+        state = state * 747796405u + 2891336453u;
+    }
+
+    uint32_t state;
+};
+
+struct PCG32
+{
+    static PCG32 construct(NBL_CONST_REF_ARG(uint32_t) initialState)
+    {
+        PCGStateHolder stateHolder = {initialState};
+        return PCG32(stateHolder);
+    }
+
+    uint32_t operator()()
+    {
+        const uint32_t word = ((stateHolder.state >> ((stateHolder.state >> 28u) + 4u)) ^ stateHolder.state) * 277803737u;
+        const uint32_t result = (word >> 22u) ^ word;
+        stateHolder.pcg32_state_advance();
+
+        return result;
+    }
+
+    PCGStateHolder stateHolder;
+};
+
+struct PCG32x2
+{
+    static PCG32x2 construct(NBL_CONST_REF_ARG(uint32_t) initialState)
+    {
+        PCG32 rng = PCG32::construct(initialState);
+        return PCG32x2(rng);
+    }
+
+    uint32_t2 operator()()
+    {
+        return uint32_t2(rng(), rng());
+    }
+
+    PCG32 rng;
+};
```

Comment:
I already merged a PCG from #803 how does this play with that?
https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/include/nbl/builtin/hlsl/random/pcg.hlsl

also the thing I merged seems to advance the state in a different way ?

Reply:
I think it came from #820. Looks mostly the same. Not sure what's going on with advancing state. All sources and examples for pcg I could find always does `state = state * multiplier + increment`.

Reply:
See: https://github.com/imneme/pcg-c-basic/blob/bc39cd76ac3d541e618606bcc6e1e5ba5e5e6aa3/pcg_basic.c#L63

Reply:
can you make a templated `PCG<IntegerScalarState>` which is specialized differently for `uint32_t` and `uint64_t` ?